Fixes discrepancy between solutions for Eq and Expr in nonlinsolve (converts Eq to Expr in _separate_poly_nonpoly)

alijosephine opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Related to issue #19144 
Fixes remaining bug from #19144 


#### Brief description of what is fixed or changed
Previously, Expr and Eq representations of same equations would give different solutions in nonlinsolve() because Expr and Eq were treated differently by _separate_poly_nonpoly(). In this commit, added Eq to Expr conversion based on 'LHS - RHS = 0' in _separate_poly_nonpoly(). Test cases have been added to test_solveset.py


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * nonlinsolve properly handles instances of Eq in systems of equations
<!-- END RELEASE NOTES -->

alijosephine committed:
solvers: converted Equality to Expr before processing eq in _separate_poly_nonpoly()

Previously, Expr and Eq representations of same equations would give different solutions in nonlinsolve() because Expr and Eq were treated differently by _separate_poly_nonpoly(). In this commit, added Eq to Expr conversion based on 'LHS - RHS = 0' in _separate_poly_nonpoly().

alijosephine committed:
solvers: test cases for issue 19144

Converted Equality to Expr before processing eq in _separate_poly_nonpoly() as part of commit a3a6b2b1afcf6449be0233a7787658a9c9e182c3. In this commit, test cases added to test_solveset.py for the same.

oscarbenjamin commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2895,3 +2895,21 @@ def test_issue_21236():
+    soln_eq = nonlinsolve(eq,[x,y])
+    assert soln_eq == soln_expr
+
+def test_issue_19144_case2():
```
There should always be two blank lines between test functions.

oscarbenjamin replied:
These two functions can just be combined.

alijosephine replied:
makes sense, will make the change

oscarbenjamin commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2895,3 +2895,21 @@ def test_issue_21236():
+    eq = [Eq(exp, 0) for exp in expr]
+    soln_expr = nonlinsolve(expr,[x,y])
+    soln_eq = nonlinsolve(eq,[x,y])
+    assert soln_eq == soln_expr
```
Better to list the actual solution explicitly:
```
soln = {(1 - I, I), (1 + I, -I)}
assert soln_eq == soln_expr == soln
```

alijosephine replied:
have committed the review comment changes

alijosephine committed:
solvers: review comments changes for unit tests 

Made changes suggested in review comments for unit tests. These tests are to verify the bugfix related to issue #19144.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -3445,6 +3445,9 @@ def _separate_poly_nonpoly(system, symbols):
     denominators = set()
     poly = None
     for eq in system:
+        # Convert equality to expression
```
This has to come after the denominators are identified, otherwise two terms with denominators might cancel, e.g. `Eq(2*x + 1/x, x + 1/x)` does not have a solution of 0.  (And that can be added as a test.)

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -3445,6 +3445,9 @@ def _separate_poly_nonpoly(system, symbols):
     denominators = set()
     poly = None
     for eq in system:
+        # Convert equality to expression
+        if isinstance(eq, Equality):
+            eq = eq.lhs - eq.rhs
         # Store denom expression if it contains symbol
         denominators.update(_simple_dens(eq, symbols))
```
```suggestion
        # Store denom expressions that contain symbols
        denominators.update(_simple_dens(eq, symbols))
        # Convert equality to expression
        if isinstance(eq, Equality):
            eq = eq.rewrite(Add)
```

alijosephine replied:
Right, had missed this case. Thank you.

smichr commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2895,3 +2895,22 @@ def test_issue_21236():
+
+
+def test_issue_19144():
+    x = Symbol('x')
+    y = Symbol('y')
+    # test case 1
+    expr1 = [x + y - 1, y**2 + 1]
+    eq1 = [Eq(exp, 0) for exp in expr1]
+    soln1 = {(1 - I, I), (1 + I, -I)}
+    soln_expr1 = nonlinsolve(expr1,[x,y])
+    soln_eq1 = nonlinsolve(eq1,[x,y])
+    assert soln_eq1 == soln_expr1 == soln1
+    # test case 2 - with denoms
+    expr2 = [x/y - 1, y**2 + 1]
+    eq2 = [Eq(exp, 0) for exp in expr2]
+    soln2 = {(-I, -I), (I, I)}
+    soln_expr2 = nonlinsolve(expr2,[x,y])
+    soln_eq2 = nonlinsolve(eq2,[x,y])
+    assert soln_eq2 == soln_expr2 == soln2
```
```suggestion
    # test case 1
    expr1 = [x + y - 1, y**2 + 1]
    eq1 = [Eq(i, 0) for i in expr1]
    soln1 = {(1 - I, I), (1 + I, -I)}
    soln_expr1 = nonlinsolve(expr1, [x, y])
    soln_eq1 = nonlinsolve(eq1, [x, y])
    assert soln_eq1 == soln_expr1 == soln1
    # test case 2 - with denoms
    expr2 = [x/y - 1, y**2 + 1]
    eq2 = [Eq(i, 0) for i in expr2]
    soln2 = {(-I, -I), (I, I)}
    soln_expr2 = nonlinsolve(expr2, [x, y])
    soln_eq2 = nonlinsolve(eq2, [x, y])
    assert soln_eq2 == soln_expr2 == soln2
    # denominators that cancel in expression
    assert nonlinsolve([Eq(x + 1/x, 1/x)], [x]) == FiniteSet((S.EmptySet,))
```

smichr committed:
Apply suggestions from code review

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -3445,8 +3445,11 @@ def _separate_poly_nonpoly(system, symbols):
         denominators.update(_simple_dens(eq, symbols))
+        # Convert equality to expression
+        if isinstance(eq, Equality):
+            eq = eq.rewrite(Add)
```
I've seen this in a few places in the codebase. Why is `rewrite(Add)` preferred over `lhs - rhs`?

To me rewriting an equation as an expression is not even really a "rewrite" since usually rewrite in sympy returns a symbolic object that is equivalent to the original.

smichr replied:
It's just a guaranteed way to know you have done the right thing: `eq.rewrite(Add)` vs `eq.rhs - eq.rhs`.

smichr replied:
(Historically, the option to write `eq.as_expr()` was frowned upon so the more specific `rewrite as an Add` was added.)

smichr committed:
Merge branch 'master' into 19144_expr_vs_eq_nonlinsolve_patch
